### PR TITLE
Optimize Redaction detection when capturing values

### PR DESCRIPTION
**Pull request by jpbempel**
# What Does This Do
Perform redaction test at instrumentation time instead at runtime

# Motivation
In some cases add significant overhead

# Additional Notes

Jira ticket: [DEBUG-2335]

[DEBUG-2335]: https://datadoghq.atlassian.net/browse/DEBUG-2335?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by cimi — COMMENTED** [review]
I need to review `CapturedContextInstrumentor` carefully, the logic there is complicated - in the meantime, can we test this somehow?

**Comment by jpbempel** [comment]
> I need to review `CapturedContextInstrumentor` carefully, the logic there is complicated - in the meantime, can we test this somehow?

I have tested it locally on petclinic and profiled it

**Comment by cimi** [comment]
I was thinking about automated testing for correctness of the redaction, not performance. I expected some test updates given we've changed the instrumentation logic.

Do we have any unit tests for the logic in `CapturedContextInstrumentor.java`? 

Or can we add a test case in the [smoke tests](https://github.com/DataDog/dd-trace-java/tree/master/dd-smoke-tests/debugger-integration-tests/src/test/java/datadog/smoketest)

Or [in the captured snapshots](https://github.com/DataDog/dd-trace-java/tree/a481d0b8c2ba49614d669bbdea07e068f4999f0e/dd-java-agent/agent-debugger/src/test/resources/com/datadog/debugger)?

(If we already have integration tests, that's good - they should cover this change - I haven't read the tests to see).

**Comment by jpbempel** [comment]
> I was thinking about automated testing for correctness of the redaction, not performance. I expected some test updates given we've changed the instrumentation logic.
> 
> Do we have any unit tests for the logic in `CapturedContextInstrumentor.java`?
> 
> Or can we add a test case in the [smoke tests](https://github.com/DataDog/dd-trace-java/tree/master/dd-smoke-tests/debugger-integration-tests/src/test/java/datadog/smoketest)
> 
> Or [in the captured snapshots](https://github.com/DataDog/dd-trace-java/tree/a481d0b8c2ba49614d669bbdea07e068f4999f0e/dd-java-agent/agent-debugger/src/test/resources/com/datadog/debugger)?
> 
> (If we already have integration tests, that's good - they should cover this change - I haven't read the tests to see).

We have already tests for in 

- [keywordRedaction CapturedSnapshotTest](https://github.com/DataDog/dd-trace-java/blob/734e3c59980f8b62181f48e8eaf72b2aeb1cde51/dd-java-agent/agent-debugger/src/test/java/com/datadog/debugger/agent/CapturedSnapshotTest.java#L1821)
- [keywordRedactionConditions CapturedSnapshotTest](https://github.com/DataDog/dd-trace-java/blob/734e3c59980f8b62181f48e8eaf72b2aeb1cde51/dd-java-agent/agent-debugger/src/test/java/com/datadog/debugger/agent/CapturedSnapshotTest.java#L1859)
- [keywordRedaction SpanDecorationProbeInstrumentationTest](https://github.com/DataDog/dd-trace-java/blob/2ca4cddaa677e5c070796fa545b2ce935ce53397/dd-java-agent/agent-debugger/src/test/java/com/datadog/debugger/agent/SpanDecorationProbeInstrumentationTest.java#L355)
- [keywordRedactionConditions SpanDecorationProbeInstrumentationTest](https://github.com/DataDog/dd-trace-java/blob/2ca4cddaa677e5c070796fa545b2ce935ce53397/dd-java-agent/agent-debugger/src/test/java/com/datadog/debugger/agent/SpanDecorationProbeInstrumentationTest.java#L386)
- [redacted GetMemberExpressionTest](https://github.com/DataDog/dd-trace-java/blob/aaa982e2c5fd7fb62f25ffd7fad79a9b0e4c2d49/dd-java-agent/agent-debugger/debugger-el/src/test/java/com/datadog/debugger/el/expressions/GetMemberExpressionTest.java#L72)
-

File: dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/instrumentation/CapturedContextInstrumentor.java [hunk 1]
```diff
@@ -745,6 +755,7 @@ private void collectReturnValue(AbstractInsnNode location, InsnList insnList) {
     // stack: [ret_value, capturedcontext, capturedcontext, null, type_name, ret_value]
     tryBox(returnType, insnList);
     // stack: [ret_value, capturedcontext, capturedcontext, null, type_name, ret_value]
+    // no name, no redaction
     addCapturedValueOf(insnList, limits);
```
Comment (cimi) [inline dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/instrumentation/CapturedContextInstrumentor.java:759]: We could redact based on function name, but I'm not convinced we should - anyway, out of scope for the current PR.

**Review by cimi — APPROVED** [review]
:+1: Thanks!

**Pull request merged**